chromium/rules.js

```diff
@@ -3,9 +3,8 @@
 (function(exports) {
 
 // Stubs so this runs under nodejs. They get overwritten later by util.js
-if (typeof util == 'undefined') {
-  let base = typeof window == 'undefined' ? global : window; // eslint-disable-line
-  Object.assign(base, {
+if (typeof util == 'undefined' || typeof global != 'undefined') {
```

Comment:
This still doesn't have to be that complex - given that it's just a stub, `if (typeof util === 'undefined') var util = { log: () => {} };` would be enough.

Reply:
Your code snippet is definitely satisfactory for this case!

However, I wanted to be extra defensive and make as few assumptions about the global state as possible. There are a lot of of reasons that just `util` could be undefined, but there are very few reasons both `util` would be undefined, and `global` is not undefined.

Like if someone re-arranges the files in the manifest.json, then I won't overwrite the local `util.log`.

Reply:
I just don't see why `global` matters. Even if someone rearranges files, you'll have `typeof util == 'undefined'` which means you set `log` stub here and later `util.js` overrides it with proper implementation. If you have current order, then `typeof util != 'undefined'` and you just don't override implementation. Hence `typeof util` check seems sufficient, but maybe I don't see the other edge cases you're trying to cover.

Reply:
Another part of my comment was that there is no point in defining `DBUG` etc. since it's either properly defined by `util.js` together with `log` function, or they simply don't matter and can be `undefined` when `log` is just a stub.

Reply:
@cowlicks Could you please make this last small change or, if you still don't think it's reasonable, at least merge this as-is?

Reply:
I think that `global` matters. The scenario you explain is totally correct, however it is a detail of our current implementation. `global` being present is a pretty strong indicator that we are running in a node environment, which is the only case I would want to stub `util`. This is true regardless of our implementation. Likewise, the reason `util.CONSTANTS` don't cause a problem is a detail of how it is used in this file, which might change.

Reply:
@cowlicks Better check for presence of `module` and `module.exports`.

Reply:
@cowlicks 

> global being present is a pretty strong indicator that we are running in a node environment, which is the only case I would want to stub util

I would say the opposite - in Node environment it's especially uncommon to pollute globals. When you're in Node env, a local util variable would be sufficient, which is why I suggested using it over actually putting anything into `global` namespace.

Reply:
Fixed in last commit of https://github.com/EFForg/https-everywhere/pull/12558 (so `util` will be a local stub when rules.js being used as Node.js library, and global variable when part of the extension).

Reply:
@RReverser

>I would say the opposite - in Node environment it's especially uncommon to pollute globals.

I'm not sure what you mean. I'm referring to the variable called `global` in a node environment, which is the name of the Global Object in node. In the browser it is called `window`. This is why I think `global` being present indicates a node environment. Wouldn't you agree? 

>  When you're in Node env, a local util variable would be sufficient, which is why I suggested using it over actually putting anything into global namespace.

If we stub a local `util` variable, then it would not get overwritten later when we actually load `util.js`. Defining `util` locally instead of globally would also further diverge from the behavior in the browser. Which is undesirable. 

At the end of the day, having to stub this out is a gross hack. It should be removed completely when we have better module loading.

Reply:
> If we stub a local util variable, then it would not get overwritten later when we actually load util.js.

In Node.js it anyway won't because you are using `module.exports` to export utils, so for now it's totally fine to stub as module-level variable.